### PR TITLE
Add more native map methods to PDC

### PR DESCRIPTION
**Pull request by Xander-de-Keijzer**
Closes issue #8925 
Accidentally closed the previous PR #9787 

PDC now implements these methods

    #put(Key, PersistentDataType, Object)
    #has(Key, PersistentDataType)
    #get(Key, PersistentDataType)
    #getOrDefault(Key, PersistentDataType, Object)
    #remove(Key)
    #has(Key)
    #putIfAbsent(Key, PersistentDataType, Object)
    #computeIfAbsent(Key, PersistentDataType, Function)
    #computeIfPresent(Key, PersistentDataType, BiFunction)
    #compute(Key, PersistentDataType, BiFunction)
    #getOrThrow(Key, PersistentDataType, Supplier)
    #getOrThrow(Key, PersistentDataType)
    #replace(Key, PersistentDataType, Object, Object)
    #replace(Key, PersistentDataType, Object)
    #merge(Key, PersistentDataType, Object, BiFunction)

### CODE REVIEW

File: patches/server/1039-Add-more-native-map-methods-to-PDC.patch [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
++        Preconditions.checkArgument(key != null, "The NamespacedKey key cannot be null");
++        Preconditions.checkArgument(type != null, "The provided type cannot be null");
++        Preconditions.checkArgument(value != null, "The provided value cannot be null");
++
++        put(key.toString(), type, value);
++    }
++    
++    private <T, Z> void put(String key, PersistentDataType<T, Z> type, Z value) {
++        this.customDataTags.put(key, this.registry.wrap(type.getPrimitiveType(), type.toPrimitive(value, adapterContext)));
++        // Paper end
```
Comment (Machine-Maker) [inline patches/server/1039-Add-more-native-map-methods-to-PDC.patch]: I don't think you need the "string" methods anymore because you can just call the put(Key,...) methods from the set(NamespacedKey) method. That way none of the precondition checks are duplicated.

File: patches/server/1039-Add-more-native-map-methods-to-PDC.patch [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
++
++    @Override
++    public <T, Z> boolean has(net.kyori.adventure.key.Key key, PersistentDataType<T, Z> type) {
++        Preconditions.checkArgument(key != null, "The NamespacedKey key cannot be null");
++        Preconditions.checkArgument(type != null, "The provided type cannot be null");
++        
++        return has(key.toString(), type);
++    }
++
++    private <T, Z> boolean has(String key, PersistentDataType<T, Z> type) {
```
Comment (Machine-Maker) [inline patches/server/1039-Add-more-native-map-methods-to-PDC.patch]: Same thing for this "string" method

**Review by Machine-Maker — CHANGES_REQUESTED** [review]

File: patches/server/1039-Add-more-native-map-methods-to-PDC.patch (hunk 1 above)
Comment (Xander-de-Keijzer) [reply patches/server/1039-Add-more-native-map-methods-to-PDC.patch]: Can you explain what you mean?
`
    @Override
    public <T, Z> void set(NamespacedKey key, PersistentDataType<T, Z> type, Z value) {
        Preconditions.checkArgument(key != null, "The NamespacedKey key cannot be null");
        Preconditions.checkArgument(type != null, "The provided type cannot be null");
        Preconditions.checkArgument(value != null, "The provided value cannot be null");

        this.customDataTags.put(key, this.registry.wrap(type.getPrimitiveType(), type.toPrimitive(value, adapterContext)));
    }
    
    @Override
    public <T, Z> void put(net.kyori.adventure.key.Key key, PersistentDataType<T, Z> type, Z value) {
        Preconditions.checkArgument(key != null, "The NamespacedKey key cannot be null");
        Preconditions.checkArgument(type != null, "The provided type cannot be null");
        Preconditions.checkArgument(value != null, "The provided value cannot be null");

        this.customDataTags.put(key, this.registry.wrap(type.getPrimitiveType(), type.toPrimitive(value, adapterContext)));
    }`
    Like this or?

Comment (Machine-Maker) [reply patches/server/1039-Add-more-native-map-methods-to-PDC.patch]: no, just call `this.put(key, type, value)` from `set`. `NamespacedKey` extends `Key`.

Comment (Xander-de-Keijzer) [reply patches/server/1039-Add-more-native-map-methods-to-PDC.patch]: nvm, fixed

Comment (Xander-de-Keijzer) [reply patches/server/1039-Add-more-native-map-methods-to-PDC.patch]: Wait this is wrong, has now calls itself whoops, can I just cast it to Key to make it call the other has or how should I fix that?

Comment (Machine-Maker) [reply patches/server/1039-Add-more-native-map-methods-to-PDC.patch]: Well yes, if the method name is the same, you have to cast. But for set/put its different so its not a concern.

Comment (Xander-de-Keijzer) [reply patches/server/1039-Add-more-native-map-methods-to-PDC.patch]: Okay, fixed 👍

**Comment by Xander-de-Keijzer** [comment]
Also added tests for the new PDC functions, all tests seem to pass

**Review by Machine-Maker — COMMENTED** [review]
At this point... might as well just add a getOrDefault method that takes the adventure key too since that is the only one missing.

File: patches/api/0442-Add-more-native-map-methods-to-PDC.patch [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+
+diff --git a/src/main/java/org/bukkit/persistence/PersistentDataContainer.java b/src/main/java/org/bukkit/persistence/PersistentDataContainer.java
+index 57609b7793122e135fa0c3b926500849379637b2..9d05980205aebb0f042dbaa389acee2ee38e4fdc 100644
+--- a/src/main/java/org/bukkit/persistence/PersistentDataContainer.java
++++ b/src/main/java/org/bukkit/persistence/PersistentDataContainer.java
+@@ -153,18 +153,361 @@ public interface PersistentDataContainer {
+     PersistentDataAdapterContext getAdapterContext();
+ 
+     // Paper start
++
```
Comment (Lulu13022002) [inline patches/api/0442-Add-more-native-map-methods-to-PDC.patch:16]: This new line is not needed

File: patches/api/0442-Add-more-native-map-methods-to-PDC.patch [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+-     * matching the provided key.
+-     * 
++     * matching the provided {@link NamespacedKey}.
++     *
+      * @param key the key for which existence should be checked.
+-     * 
++     *
+      * @return whether the key exists
+-     * 
++     *
```
Comment (Lulu13022002) [inline patches/api/0442-Add-more-native-map-methods-to-PDC.patch:358]: ~Theses changes are not needed (even if it's correct it's more diff)~
Nvm these change can be done in the paper patch directly: https://github.com/PaperMC/Paper/blob/master/patches/api/0336-Add-new-overload-to-PersistentDataContainer-has.patch might also merge the patches at this point

File: patches/server/1039-Add-more-native-map-methods-to-PDC.patch [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
++        return true;
++    }
++
++    @Override
++    public <T, Z> Z replace(net.kyori.adventure.key.Key key, PersistentDataType<T, Z> type, Z value) {
++        Preconditions.checkArgument(key != null, "The key cannot be null");
++        Preconditions.checkArgument(type != null, "The provided type cannot be null");
++        Preconditions.checkArgument(value != null, "The provided value cannot be null");
++        
++        Z curvalue;
```
Comment (Lulu13022002) [inline patches/server/1039-Add-more-native-map-methods-to-PDC.patch:194]: curvalue should be named curValue to use the same naming format (camel case)

File: patches/server/1039-Add-more-native-map-methods-to-PDC.patch [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
++
++import net.kyori.adventure.key.Key;
++import org.bukkit.NamespacedKey;
++import org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer;
++import org.bukkit.craftbukkit.persistence.CraftPersistentDataTypeRegistry;
++import org.bukkit.persistence.PersistentDataType;
++import org.junit.jupiter.api.Test;
++
++import java.util.NoSuchElementException;
++import java.util.function.Supplier;
```
Comment (Lulu13022002) [inline patches/server/1039-Add-more-native-map-methods-to-PDC.patch:244]: This import is not needed

File: patches/server/1039-Add-more-native-map-methods-to-PDC.patch [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
++import static org.junit.jupiter.api.Assertions.assertEquals;
++import static org.junit.jupiter.api.Assertions.assertNotEquals;
++import static org.junit.jupiter.api.Assertions.assertNull;
++import static org.junit.jupiter.api.Assertions.assertThrows;
++import static org.junit.jupiter.api.Assertions.assertTrue;
++import static org.junit.jupiter.api.Assertions.assertFalse;
++
++public class PersistentDataContainerTest {
++    
++    static private CraftPersistentDataContainer createContainer() {
```
Comment (Lulu13022002) [inline patches/server/1039-Add-more-native-map-methods-to-PDC.patch:256]: If you need to do something before each test you can use a @ BeforeEach method that would create an empty pdc for each test into a class field. Also you don't need to get a craft implementation, can use the api.

File: patches/server/1039-Add-more-native-map-methods-to-PDC.patch [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+@@ -69,6 +90,12 @@ public class CraftPersistentDataContainer implements PersistentDataContainer {
+ 
+     @Override
+     public <T, Z> Z getOrDefault(NamespacedKey key, PersistentDataType<T, Z> type, Z defaultValue) {
++        // Paper start
++        return getOrDefault((net.kyori.adventure.key.Key) key, type, defaultValue);
++    }
++    
++    @Override
++    public <T, Z> Z getOrDefault(net.kyori.adventure.key.Key key, PersistentDataType<T, Z> type, Z defaultValue) {
```
Comment (Lulu13022002) [inline patches/server/1039-Add-more-native-map-methods-to-PDC.patch:63]: The Paper end comment is missing

File: patches/server/1039-Add-more-native-map-methods-to-PDC.patch [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
++    static private CraftPersistentDataContainer createContainer() {
++        CraftPersistentDataTypeRegistry registry = new CraftPersistentDataTypeRegistry();
++        CraftPersistentDataContainer container = new CraftPersistentDataContainer(registry);
++        return container;
++    }
++    
++    @Test
++    public void basicTest() {
++        CraftPersistentDataContainer container = createContainer();
++        Key key = new NamespacedKey("test", "key");
```
Comment (Lulu13022002) [inline patches/server/1039-Add-more-native-map-methods-to-PDC.patch:265]: You can create an adventure key directly for all your cases i think

**Review by Lulu13022002 — COMMENTED** [review]
Don't forget to update the DirtyCraftPersistentDataContainer class especially now there are some methods that doesn't put the dirty flag (the adventure keys one)

File: patches/api/0442-Add-more-native-map-methods-to-PDC.patch [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
++     * This method is only usable for custom object keys. Overwriting existing
++     * tags, like the the display name, will not work as the values are stored
++     * using your namespace.
++     *
++     * @param key the key the value is stored under
++     * @param type the type which primitive storage type has to match the value
++     * @param <T> the generic type of the stored primitive
++     * @param <Z> the generic type of the eventually created complex object
++     *
++     * @return if a value
```
Comment (Lulu13022002) [inline patches/api/0442-Add-more-native-map-methods-to-PDC.patch:63]: This doesn't seems complete, it seems spigot method has the same mistake

**Comment by Xander-de-Keijzer** [comment]
> Are you planning on addressing the comments ? Otherwise I can look into taking over the PR and completing it for you.

Sorry forgot about this, I can implement the requested changes but i'm unsure of how to merge this into the Add-new-overload-to-PersistentDataContainer path so if you want to take over the PR that's fine

**Comment by Leguan16** [comment]
> but i'm unsure of how to merge this into the Add-new-overload-to-PersistentDataContainer patch

You can look at the CONTRIBUTING.md file for that. It explains it really nice how to edit existing patches.
